File: src/engine/ListDefinitionsOrigin.ts
import { InkListItem } from "./InkList";
import { ListValue } from "./Value";
import { ListDefinition } from "./ListDefinition";
import { TryGetResult } from "./TryGetResult";
import { throwNullException } from "./NullException";

export class ListDefinitionsOrigin {
  protected _lists: Map<string, ListDefinition>;
  protected _allUnambiguousListValueCache: Map<string, ListValue>;

  constructor(lists: ListDefinition[]) {
    this._lists = new Map();
    this._allUnambiguousListValueCache = new Map();

    for (let list of lists) {
      this._lists.set(list.name, list);

      for (let [key, val] of list.items) {
        let item = InkListItem.fromSerializedKey(key);
        let listValue = new ListValue(item, val);

        if (!item.itemName) {
          throw new Error("item.itemName is null or undefined.");
        }

        this._allUnambiguousListValueCache.set(item.itemName, listValue);
        this._allUnambiguousListValueCache.set(item.fullName, listValue);
      }
    }
  }
  get lists(): ListDefinition[] {
    let listOfLists: ListDefinition[] = [];

    for (let [, value] of this._lists) {
      listOfLists.push(value);
    }

    return listOfLists;
  }
  public TryListGetDefinition(
    name: string | null,
    /* out */ def: ListDefinition | null
  ): TryGetResult<ListDefinition | null> {
    if (name === null) {
      return { result: def, exists: false };
    }
    // initially, this function returns a boolean and the second parameter is an out.
    let definition = this._lists.get(name);
    if (!definition) return { result: def, exists: false };

    return { result: definition, exists: true };
  }
  public FindSingleItemListWithName(name: string | null) {
    if (name === null) {
      return throwNullException("name");
    }
    let val = this._allUnambiguousListValueCache.get(name);

    if (typeof val !== "undefined") {
      return val;
    }

    return null;
  }
}

File: ink_dart/lib/src/engine/list_definitions_origin.dart
import 'ink_list.dart';
import 'list_definition.dart';
import 'value.dart';

/// Origin container for all LIST definitions in a story.
///
/// Provides lookup and resolution of list items across all
/// defined lists in the story.
class ListDefinitionsOrigin {
  final List<ListDefinition> _lists;
  final Map<String, ListDefinition> _lookup = {};

  ListDefinitionsOrigin(this._lists) {
    for (final def in _lists) {
      _lookup[def.name] = def;
    }
  }

  /// All list definitions.
  List<ListDefinition> get lists => _lists;

  /// Try to get a list definition by name.
  ({bool exists, ListDefinition? result}) tryListGetDefinition(
    String? name,
    ListDefinition? defaultValue,
  ) {
    if (name == null) {
      return (exists: false, result: defaultValue);
    }
    final def = _lookup[name];
    if (def != null) {
      return (exists: true, result: def);
    }
    return (exists: false, result: defaultValue);
  }

  /// Find a single-item list by name.
  ListValue? findSingleItemListWithName(String name) {
    for (final def in _lists) {
      if (def.containsItemWithName(name)) {
        final item = InkListItem.fromFullName('${def.name}.$name');
        final valueResult = def.tryGetValueForItem(item);
        if (valueResult.exists) {
          final list = InkList();
          list.add(item, valueResult.result!);
          return ListValue.fromList(list);
        }
      }
    }
    return null;
  }

  @override
  String toString() => 'ListDefinitionsOrigin(${_lists.length} lists)';
}
